Microsoft.Research/Contracts/MsCorlib/System.Reflection.CustomAttributeExtensions.cs

```diff
@@ -0,0 +1,1231 @@
+    //
+    // Exceptions:
+    //   T:System.ArgumentNullException:
+    //     element is null.
+    [Pure]
+    public static IEnumerable<T> GetCustomAttributes<T>(this Assembly element) where T : Attribute
+    {
+      Contract.Requires(element != null);
+      Contract.Ensures(Contract.Result<IEnumerable<Attribute>>() != null);
+      return default(IEnumerable<T>);
+    }
+
+    //
+    // Summary:
+    //     Retrieves a collection of custom attributes of a specified type that are applied
+    //     to a specified parameter, and optionally inspects the ancestors of that parameter.
+    //
+    // Parameters:
+    //   element:
+    //     The parameter to inspect.
+    //
+    //   inherit:
+    //     true to inspect the ancestors of element; otherwise, false.
+    //
+    // Type parameters:
+    //   T:
+    //     The type of attribute to search for.
+    //
+    // Returns:
+    //     A collection of the custom attributes that are applied to element and that match
+    //     T, or an empty collection if no such attributes exist.
+    //
+    // Exceptions:
+    //   T:System.ArgumentNullException:
+    //     element is null.
+    //
+    //   T:System.NotSupportedException:
+    //     element is not a constructor, method, property, event, type, or field.
+    //
+    //   T:System.TypeLoadException:
+    //     A custom attribute type cannot be loaded.
+    [Pure]
+    public static IEnumerable<T> GetCustomAttributes<T>(this ParameterInfo element, bool inherit) where T : Attribute
+    {
+      Contract.Requires(element != null);
+      Contract.Ensures(Contract.Result<IEnumerable<Attribute>>() != null);
+      return default(IEnumerable<T>);
+    }
+
+    //
+    // Summary:
+    //     Retrieves a collection of custom attributes of a specified type that are applied
+    //     to a specified member, and optionally inspects the ancestors of that member.
+    //
+    // Parameters:
+    //   element:
+    //     The member to inspect.
+    //
+    //   inherit:
+    //     true to inspect the ancestors of element; otherwise, false.
+    //
+    // Type parameters:
+    //   T:
+    //     The type of attribute to search for.
+    //
+    // Returns:
+    //     A collection of the custom attributes that are applied to element and that match
+    //     T, or an empty collection if no such attributes exist.
+    //
+    // Exceptions:
+    //   T:System.ArgumentNullException:
+    //     element is null.
+    //
+    //   T:System.NotSupportedException:
+    //     element is not a constructor, method, property, event, type, or field.
+    //
+    //   T:System.TypeLoadException:
+    //     A custom attribute type cannot be loaded.
+    [Pure]
+    public static IEnumerable<T> GetCustomAttributes<T>(this MemberInfo element, bool inherit) where T : Attribute
+    {
+      Contract.Requires(element != null);
+      Contract.Ensures(Contract.Result<IEnumerable<Attribute>>() != null);
```

Comment:
:question: Any reason not to make this use `IEnumerable<T>` instead of `IEnumerable<Attribute>`? My guess is the current form was easier for copy/paste. I'm assuming the static checker doesn't have a problem with variance like this?